language/move-prover/src/prelude.bpl

```diff
@@ -546,156 +562,129 @@ procedure {:inline 1} $InitVerification() {
 }
 
 // ============================================================================================
-// Specifications
+// Functional APIs
 
 // TODO: unify some of this with instruction procedures to avoid duplication
 
 // Tests whether resource exists.
-function {:inline} $ResourceExistsRaw(m: $Memory, resource: $TypeValue, addr: int): bool {
-    domain#$Memory(m)[$Global(resource, addr)]
+function {:inline} $ResourceExistsRaw(m: $Memory, args: $TypeValueArray, addr: int): bool {
+    domain#$Memory(m)[args, addr]
 }
-function {:inline} $ResourceExists(m: $Memory, resource: $TypeValue, address: $Value): $Value {
-    $Boolean($ResourceExistsRaw(m, resource, a#$Address(address)))
+function {:inline} $ResourceExists(m: $Memory, args: $TypeValueArray, addr: $Value): $Value {
+    $Boolean($ResourceExistsRaw(m, args, a#$Address(addr)))
 }
 
 // Obtains Value of given resource.
-function {:inline} $ResourceValue(m: $Memory, resource: $TypeValue, address: $Value): $Value {
-  contents#$Memory(m)[$Global(resource, a#$Address(address))]
+function {:inline} $ResourceValue(m: $Memory, args: $TypeValueArray, addr: $Value): $Value {
+  contents#$Memory(m)[args, a#$Address(addr)]
 }
 
 // Applies a field selection to a Value.
-function {:inline} $SelectField(val: $Value, field: $FieldName): $Value { //breaks abstracts, we don't know $Fieldname = int
+function {:inline} $SelectField(val: $Value, field: $FieldName): $Value {
     $select_vector(val, field)
 }
 
 // Dereferences a reference.
-function {:inline} $Dereference(ref: $Reference): $Value {
-    v#$Reference(ref)
+function {:inline} $Dereference(ref: $Mutation): $Value {
+    v#$Mutation(ref)
 }
 
-// Check whether sender account exists.
-function {:inline} $ExistsTxnSenderAccount(m: $Memory, txn: $Transaction): bool {
-   domain#$Memory(m)[$Global($LibraAccount_T_type_value(), sender#$Transaction(txn))]
-}
-
-function {:inline} $TxnSender(txn: $Transaction): $Value {
-    $Address(sender#$Transaction(txn))
-}
-
-// Forward declaration of type Value of LibraAccount. This is declared so we can define
-// $ExistsTxnSenderAccount and $LibraAccount_save_account
-const unique $LibraAccount_T: $TypeName;
-function $LibraAccount_T_type_value(): $TypeValue;
-axiom is#$StructType($LibraAccount_T_type_value()) && name#$StructType($LibraAccount_T_type_value()) == $LibraAccount_T;
-function $LibraAccount_Balance_type_value(tv: $TypeValue): $TypeValue;
-
 // ============================================================================================
 // Instructions
 
-procedure {:inline 1} $Exists(address: $Value, t: $TypeValue) returns (dst: $Value)
-{{backend.type_requires}} is#$Address(address);
-{
-    dst := $ResourceExists($m, t, address);
-}
-
-procedure {:inline 1} $MoveToRaw(ta: $TypeValue, a: int, v: $Value)
+procedure {:inline 1} $MoveToRaw(m: $Memory, ta: $TypeValueArray, a: int, v: $Value) returns (m': $Memory)
```

Comment:
Since all of global memory is being treated functionally (as inout parameter essentially), do we not even need global declarations ($m) any more?

Reply:
It's removed. We do not have a single state variable in the prelude, $txn is also gone.